src/devices/video/voodoo_render.cpp

```diff
@@ -0,0 +1,2957 @@
+	{
+		int color = (val >> 0) & 0xff;
+		int g = (val >> 8) & 1;
+		int x = (val >> 9) & 3;
+		int y = (val >> 11) & 3;
+
+		if (!g)
+		{
+			s_dither4_lookup[val] = dither_rb(color, s_dither_matrix_4x4[y * 4 + x]);
+			s_dither2_lookup[val] = dither_rb(color, s_dither_matrix_2x2[y * 4 + x]);
+		}
+		else
+		{
+			s_dither4_lookup[val] = dither_g(color, s_dither_matrix_4x4[y * 4 + x]);
+			s_dither2_lookup[val] = dither_g(color, s_dither_matrix_2x2[y * 4 + x]);
+		}
+	}
+	for (int val = 0; val < 256*4*2; val++)
+	{
+		int color = (val >> 0) & 0xff;
+		int g = (val >> 8) & 1;
+
+		if (!g)
+			s_nodither_lookup[val] = color >> 3;
+		else
+			s_nodither_lookup[val] = color >> 2;
+	}
+}
+
+
+
+//**************************************************************************
+//  STW HELPER
+//**************************************************************************
+
+// use SSE on 64-bit implementations, where it can be assumed
+#if 1 && ((!defined(MAME_DEBUG) || defined(__OPTIMIZE__)) && (defined(__SSE2__) || defined(_MSC_VER)) && defined(PTR64))
+
+#include <emmintrin.h>
+#ifdef __SSE4_1__
+#include <smmintrin.h>
+#endif
+
+class stw_helper
+{
+public:
+	stw_helper() { }
+	stw_helper(stw_helper const &other) = default;
+	stw_helper &operator=(stw_helper const &other) = default;
+
+	void set(s64 s, s64 t, s64 w)
+	{
+		m_st = _mm_set_pd(s << 8, t << 8);
+		m_w = _mm_set1_pd(w);
+	}
+
+	bool is_w_neg() const
+	{
+		return _mm_comilt_sd(m_w, _mm_set1_pd(0.0));
+	}
+
+	void get_st_shiftr(s32 &s, s32 &t, s32 shift) const
+	{
+		shift += 8;
+		s = _mm_cvtsd_si64(_mm_shuffle_pd(m_st, _mm_setzero_pd(), 1)) >> shift;
+		t = _mm_cvtsd_si64(m_st) >> shift;
+	}
+
+	void add(stw_helper const &delta)
+	{
+		m_st = _mm_add_pd(m_st, delta.m_st);
+		m_w = _mm_add_pd(m_w, delta.m_w);
+	}
+
+	void calc_stow(s32 &sow, s32 &tow, s32 &oowlog) const
+	{
+		__m128d tmp = _mm_div_pd(m_st, m_w);
+		__m128i tmp2 = _mm_cvttpd_epi32(tmp);
+#ifdef __SSE4_1__
+		sow = _mm_extract_epi32(tmp2, 1);
+		tow = _mm_extract_epi32(tmp2, 0);
+#else
+		sow = _mm_cvtsi128_si32(_mm_shuffle_epi32(tmp2, _MM_SHUFFLE(0, 0, 0, 1)));
+		tow = _mm_cvtsi128_si32(tmp2);
+#endif
+		oowlog = -fast_log2(_mm_cvtsd_f64(m_w), 0);
+	}
+
+private:
+	__m128d m_st;
+	__m128d m_w;
+};
+
+#else
+
+class stw_helper
+{
+public:
+	stw_helper() {}
+	stw_helper(stw_helper const &other) = default;
+	stw_helper &operator=(stw_helper const &other) = default;
+
+	void set(s64 s, s64 t, s64 w)
+	{
+		m_s = s;
+		m_t = t;
+		m_w = w;
+	}
+
+	bool is_w_neg() const
+	{
+		return (m_w < 0) ? true : false;
+	}
+
+	void get_st_shiftr(s32 &s, s32 &t, s32 shift) const
+	{
+		s = m_s >> shift;
+		t = m_t >> shift;
+	}
+
+	inline void add(stw_helper const &other)
+	{
+		m_s += other.m_s;
+		m_t += other.m_t;
+		m_w += other.m_w;
+	}
+
+	// Computes s/w and t/w and returns log2 of 1/w
+	// s, t and c are 16.32 values.  The results are 24.8.
+	inline void calc_stow(s32 &sow, s32 &tow, s32 &oowlog) const
```

Comment:
`inline` is implied when the body is present at declaration for member functions.

Reply:
Removed inline here.